SQL client support (#17173)

devozerov opened the pull request:
This PR introduces client-side support for SQL. The design is explained in the issue https://github.com/hazelcast/hazelcast/issues/17173#issue-650589250

Ent PR: https://github.com/hazelcast/hazelcast-enterprise/pull/3666
Protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/329

Closes #17173

viliam-durina commented on hazelcast/src/main/java/com/hazelcast/client/LoadBalancer.java:
```diff
@@ -43,9 +43,10 @@
+     * @return Returns the next data member or null if no member is available
      */
-    Member next();
+    Member next(boolean dataMember);
```
We can't change public API.

devozerov replied:
Fixed. Please take a look

viliam-durina commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlPage.java:
```diff
@@ -0,0 +1,68 @@
+            return false;
+        }
+
+        return rows != null ? rows.equals(page.rows) : page.rows == null;
```
Suggested by intellij
```suggestion
        return Objects.equals(rows, page.rows);
```

devozerov replied:
Fixed

viliam-durina commented on hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:
```diff
@@ -767,8 +779,25 @@ public ClientConnection getRandomConnection() {
+                Member member = client.getClientClusterService().getMember(memberId);
+
+                if (member == null || member.isLiteMember()) {
+                    return null;
```
This is a fallback code that's used when the load balancer returned `null`, even though it should return null only if there's no member available. But it's also a fallback when the load balancer doesn't implement the `nextDataMember` method, and in this case we should find the first data member. 

What if the first member returned by the iterator isn't a data member? A user upgrading to 4.2 that has a custom LB can get random "Client must be connected to at least one data member to execute SQL queries" errors.

viliam-durina replied:
It's also used for non-smart client - can such client be connected to a lite member? Won't SQL work at all in that case?

devozerov replied:
Not really. LB is only used when smart routing is enabled (see `isSmartRoutingEnabled` check). Otherwise, we always use the very first member connection (aka "unisocket" client). 

@sancar could you please give your comments on the correct behavior?

devozerov replied:
Reworked the connection handling so that `dataMember=true` will return a connection to a data member, even if it is not the first one.

devozerov commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/DefaultMessageTaskFactoryProvider.java:
```diff
@@ -1779,6 +1785,15 @@ private void initializeManagementCenterTaskFactories() {
+            (cm, con) -> new SqlFetchMessageTask(cm, node, con));
+        factories.put(SqlCloseCodec.REQUEST_MESSAGE_TYPE,
+            (cm, con) -> new SqlCloseMessageTask(cm, node, con));
+//endregion
```
Fixed

devozerov commented on hazelcast/src/main/java/com/hazelcast/client/util/StaticLB.java:
```diff
@@ -38,7 +38,7 @@ public void init(Cluster cluster, ClientConfig config) {
-    public Member next() {
-        return member;
+    public Member next(boolean dataMember) {
+        return dataMember && member.isLiteMember() ? null : member;
```
I am not sure I understand what you mean. This is a static load balancer that always returns the same member. If this member doesn't happen to be a data member, we return `null`.

viliam-durina replied:
I think it would be better to throw in this case. By setting the `StaticLB` one asks that all traffic goes to a specific member, but here we return null which in turn will go to some data member.

devozerov replied:
This is OK, since traffic may go to another member for many reasons:
1. If smart routing is disabled
1. If the member that is set is no longer reachable
1. Now we add the third case - if the added member is not a data member

devozerov replied:
In other words, there is no such semantics, that forces us to use a specific member. Load balancer is a hint.

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/QueryException.java:
```diff
@@ -62,15 +63,19 @@ public static QueryException error(int code, String message, Throwable cause, UU
+    }
+
+    public static QueryException clientMemberConnection(UUID clientId) {
+        return error(SqlErrorCode.MEMBER_CONNECTION, "Client member cannot be reached: " + clientId);
```
Fixed

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/NodeServiceProviderImpl.java:
```diff
@@ -63,6 +64,17 @@ public UUID getLocalMemberId() {
     }
 
+    @Override
+    public Collection<UUID> getClientMembersIds() {
```
Changed to Set

devozerov commented on hazelcast/src/main/java/com/hazelcast/client/util/AbstractLoadBalancer.java:
```diff
@@ -32,8 +34,9 @@
-    private final AtomicReference<Member[]> membersRef = new AtomicReference<Member[]>(new Member[0]);
+    private static final Member[] EMPTY_MEMBERS = new Member[0];
 
+    private final AtomicReference<Members> membersRef = new AtomicReference<>(new Members(EMPTY_MEMBERS, EMPTY_MEMBERS));
```
Agree, but as you mentioned, this is not part of this PR

viliam-durina commented on hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryStateRegistryUpdater.java:
```diff
@@ -153,6 +160,12 @@ private void checkMemberState() {
         }
 
+        private void checkClientState() {
+            Collection<UUID> activeClientIds = nodeServiceProvider.getClientIds();
```
Should be `Set<UUID>` also here

devozerov replied:
Fixed, now Set is used.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:
```diff
@@ -0,0 +1,218 @@
+                    params0.add(serializeParameter(param));
+                }
+            } else {
+                params0 = null;
```
It would be much easier to understand if we represent the empty list with an actually empty list rather than `null`. 
I have checked the `SqlExecuteMessageTask` code. It would simplify the code there also. 
Instead of 
```
 if (parameters.parameters != null && !parameters.parameters.isEmpty()) {
                for (Data param : parameters.parameters) {
                    query.addParameter(serializationService.toObject(param));
                }
            }
```
the following will suffice
```
                for (Data param : parameters.parameters) {
                    query.addParameter(serializationService.toObject(param));
                }      
```

devozerov replied:
Fixed, now parameters are non-null.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:
```diff
@@ -0,0 +1,218 @@
+
+        ClientInvocationFuture fut = invocation.invoke();
+
+        return new ClientDelegatingFuture<T>(fut, getSerializationService(), decoder, false).get();
```
You don't need the ClientDelegatingFuture since you are doing a sync call. It is mostly needed when we need to pass a future to the user(or other part of the system)

Why not just do the following instead? 
```
ClientMessage clientMessage = fut.get();
return decoder.decodeClientMessage(clientMessage);
```

Even further, there is no need for static decoder functions because they are all sync calls. 
I would suggest making the return type `ClientMessage`.  For the fetch method: 
```
            ClientMessage responseMessage = invoke(message, connection);
            SqlFetchCodec.ResponseParameters response = SqlFetchCodec.decodeResponse(responseMessage);
            SqlFetchResponse res = new SqlFetchResponse(response.rowPage, response.error);
            handleResponseError(res.getError());
```

Static decoder functions and delegating future are used for async cases. There is no need to complicate the code with them when not needed.

devozerov replied:
Simplified as you suggested.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:
```diff
@@ -0,0 +1,218 @@
+        if (connection == null) {
+            throw rethrow(QueryException.error(
+                SqlErrorCode.CONNECTION_PROBLEM,
+                "Client must be connected to at least one data member to execute SQL queries"
```
So a unısocket client that is connected to a lite member can not run a query? 
This is not suited to our unisocket client behavior. Our unisocket clients should be able to operate same regardless of the member type they are connected to.

sancar replied:
You can check `CreateProxyMessageTask`. It is one of the places that we explicitly redirect to a data member. 

The partition-based ones solve the problem automatically since they will run on their partition owner anyway. Even if we invoke the operation on a lite member, the invocation system redirects it to the partition owner.

devozerov replied:
Yes, SQL may not work if there lite members in the topology. This is ok for the beta stage at least. Lite member support might be added in the future if we see a demand from users. 
This limitation will be mentioned explicitly in the product's documentation.

viliam-durina replied:
I think it's pretty bad that it will fail to work randomly, depending on the member the non-smart client happens to connect. Users might try it and see that it works and then have this misleading error at runtime. In that case we should disallow non-smart clients entirely.

devozerov replied:
Documentation solves this. Again - we do not support lite members not because we are lazy, or because we do not understand the implications. This is a necessary measure given the huge scope. 
Moreover, in the current PR, we iterate over all members to find the data member for SQL. The flakiness you mentioned is only possible if a user will flip-flop member state from lite member and vice versa. This is extremely unlikely.

sancar replied:
Iterating all connections does not help unisocket client. It will only have one connection. If the client disconnect from data member and connects to lite member, then it will fail. 

If this is the decision, we basically say that for the uniscoket client, if you use query, your cluster should not contain a lite member. For the smart client, I agree that the problem is very unlikely to occur.

asimarslan commented on hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java:
```diff
@@ -35,20 +35,20 @@
         this.memberIdLow = memberIdLow;
-        this.localHigh = localHigh;
-        this.localLow = localLow;
+        this.localIdHigh = localIdHigh;
```
Both serialization service and client protocol have built-in UUID support. If the only purpose of these High and Low part splits is to support them, I think you can just use UUID instead.

asimarslan replied:
protocol schema will redirect you if you configure it on the protocol side.
https://github.com/hazelcast/hazelcast-client-protocol/blob/master/schema/protocol-schema.json#L26

devozerov replied:
`UUID` content is inlined into the `QueryId` to reduce the amount of litter. Real UUIDs are rarely extracted from this object. We only do `equals/hashCode` on hot paths.

sancar replied:
@devozerov I understand the optimization but UUID is a common class that we use. This will be an exception to how we represent it. I would prefer using UUID for the sake of readability.

devozerov replied:
Let me highlight this once again - `QueryId` is not `UUID`. We use two `UUID`-s to guarantee its uniqueness, but other than that it is mostly an opaque blob.

asimarslan replied:
This way of this UUID usage does not match to protocol style. Not all languages use to high low split of a UUID as java.
MemberID is already created somewhere on the member, you don't need to split it for litter optimization. let the protocol encode/Decode it.
For the `localId`, how many local id's are created in the system? if they are per query, I don't think you should worry about the litter it creates. 
We already use UUID in memberId and listenerRegistrationId etc, which are also infrequent.
I understand you want to optimize it in the best way possible but protocol definition is also an important part of our architecture. Moreover, we cannot change the protocol when we publish it.

devozerov replied:
Colleagues, I would like to emphasize one more time what is the nature of query ID. From the client perspective, it is an opaque blob that is passed back and forth. From the server perspective, this is an object that serves multiple purposes, such as 1) guaranteeing cluster-wide uniqueness; 2) providing an ability to find the originator member ID without remote requests. 

The design of the client protocol cannot enforce the content of this object, because client doesn't use any features of it, while the server does. Therefore, it cannot be `UUID`.

On the other hand, I share your concerns about future protocol changes. So I propose the other solution - to change it to `String`. This would be a good choice from the protocol perspective for an opaque object. If we would like to change the content of this object on the server, the client protocol will not be affected. 

I merged these changes to the PR, please take a look.

asimarslan replied:
@devozerov if the client doesn't need the content of this, I think 4 long value representation is much better than string. String needs to be parsed on server side and it's not a good idea. 
Our initial concern was about parsing the internal UUIDs but as you say the client won't need it.
The original QueryId should be used instead of a string.

devozerov replied:
I used to think that QueryId was good solution, but now I think that String is better.

As explained before, the query ID is completely opaque to the client, I.e. it is only passed back and forth. String representation has two advantages over QueryId:
1. It is more stable - we can change the server-side representation without breaking protocol compatibility
1. It makes client implementation simpler. Consider that we will have SQL implemented in many clients in future. All of them would have to implement a custom object that does zero useful work

That said, IMO String is good enough. Also please note that there is no parsing in the sever side. We convert QueryId to String once (see QueryId.unparse) and them use this String as a key. There should be no negative performance impact from this. 

WDYT?

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:
```diff
@@ -0,0 +1,167 @@
+        while (iterator.hasNext()) {
+            SqlRow row = iterator.next();
+            Row rowInternal = ((SqlRowImpl) row).getDelegate();
+            Data rowData = serializationService.toData(rowInternal);
```
This does not seem correct. We are not using Data for builtin `IdentifiedDataSerializable` objects on our protocol as much as possible. It is hard to define what is Data in the protocol. More importantly, they don't support evolution, it is very likely to make a mistake. 

I see that `Row` is an interface. Which `Row` implementations can we return to the client?

devozerov replied:
I'll change it to `List<Data>`. There is no model for a row at the moment.

devozerov replied:
Fixed. Now `List<Data>` is used to represent row values. Please take a look.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:
```diff
@@ -0,0 +1,167 @@
+ */
+public class QueryClientStateRegistry {
+    /** Registered client cursors. */
+    private final ConcurrentHashMap<QueryId, QueryClientState> clientCursors = new ConcurrentHashMap<>();
```
I could not find the cleanup code for the case where a client lefts the cluster without closing the query.
You can use `ClientEndpoint.addDestroyAction(UUID registrationId, Callable<Boolean> removeAction);` for this. 

A client enndpoint is removed and all desrtoy actions are called when the assocaiated connection is disconnected. Since the queries are desinged to be `sticky` it looks like a good fit.

sancar replied:
And can we have a test to verify that cleanups are done correctly?

sancar replied:
I see that QueryClientStateRegistry.update is used for this purpose. I think, our standard solution with ClientEndpoint addDestroyAction is enough for this. What do you think ?

devozerov replied:
Please see:
com.hazelcast.sql.impl.client.QueryClientStateRegistry#update
com.hazelcast.sql.SqlErrorClientTest#testCursorCleanupOnClientLeave

devozerov replied:
`ClientEndpoint addDestroyAction` looks heavier to me - one have to register it for every request, and also be careful to deregister it properly. 
The key observation is that majority of requests will be completed normally, without client disconnects. I think the current approach handles it better - we do no do any additional actions for the most queries.

sancar replied:
Ok. Fair enough.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlExecuteMessageTask.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Override
+    public Permission getRequiredPermission() {
+        return null;
```
Is permission left `null` on purpose? I think, SQL permissions are important.

devozerov replied:
We do not have a security model yet. Given that an SQL script could trigger any kind of action (e.g. read from several maps, or start Jet job, etc), it is impossible to define the required permissions on that level. The permissions will be known after parsing only.
Meanwhile, we may add a global flag to disable SQL on the member, but this is out of scope of this PR.

devozerov replied:
I think we may define a global "query" permission that will govern whether the user can execute any query or not. And more deep integration will be implemented later.

devozerov replied:
Added the `sql` permission, modeled after the `transaction` permission.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlExecuteMessageTask.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Override
+    public String getMethodName() {
+        return "execute";
```
for the method name, we usually use the user API. And note that related tests are on the enterprise repo. Make sure that you have `SecurityInterceptor` and permission tests for the new use case.

devozerov replied:
I can change it to `query`, but what is the goal of this? From the user perspective, he executes a query, but internally it is translated into several different commands, that are not directly mapped to the public API. Moreover, the result of this operation contains internal objects, which also highlights that this is not a direct API call.

sancar replied:
Well, we don't really have a clear answer. In the end, whatever you choose must be documented clearly. We select it as same as the public method name to make it more intuitive. Not sure what to do when it is multiple requests per API call. 
Just be aware that what you choose here is also a public API and needs to be documented.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1366/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1366/Hazelcast-pr-builder-1366.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CPMemberAddRemoveTest.when_snapshotIsTakenWhileRemovingCPMember_newMemberInstallsSnapshot:770->HazelcastTestSupport.assertTrueEventually:1344->HazelcastTestSupport.assertTrueEventually:1242->lambda$when_snapshotIsTakenWhileRemovingCPMember_newMemberInstallsSnapshot$17:773 expected:<[CPMember{uuid=8760c6d3-885d-405b-be50-eb7141f8ac92, address=[127.0.0.1]:5702}, CPMember{uuid=88ad5201-619c-415a-a203-1e1b08e3a8b3, address=[127.0.0.1]:5703}, CPMember{uuid=841abf46-81f8-4a95-a02e-0fb709b9443b, address=[127.0.0.1]:5704}]> but was:<[CPMember{uuid=8760c6d3-885d-405b-be50-eb7141f8ac92, address=[127.0.0.1]:5702}, CPMember{uuid=88ad5201-619c-415a-a203-1e1b08e3a8b3, address=[127.0.0.1]:5703}, CPMember{uuid=9bd91814-9ff9-42a7-95c3-e3c57981e016, address=[127.0.0.1]:5701}]>
[INFO] 
[ERROR] Tests run: 31141, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:
```diff
@@ -0,0 +1,176 @@
+        List<Data> values = new ArrayList<>(columnCount);
+
+        for (int i = 0; i < columnCount; i++) {
+            values.add(serializationService.toData(row.getObject(i)));
```
Not sure that I got why do we need to convert the result to `Data` if we are already on the client?

devozerov replied:
This code is invoked on the member

petrpleshachkov replied:
Shall we move it then somewhere to the member related class?

devozerov replied:
Makes sense. Moved to the `com.hazelcast.sql.impl.state` package.

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java:
```diff
@@ -0,0 +1,181 @@
+                service.close(connection, queryId);
+            }
+        } finally {
+            closed = true;
```
If an exception is thrown from` service.close(connection, queryId)` then we still mark the iterator as closed while it is not?

devozerov replied:
The only reason for the exception here is a broken connection. If it is broken, there is no sense trying to keep the result opened, because no more rows could be fetched anyway.

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    public void setCursorBufferSize(int cursorBufferSize) {
+        assert cursorBufferSize >= 0;
```
What is the meaning for the `cursorBufferSize=0`?

devozerov replied:
These methods are not used in this PR, they are needed by JDBC. Removed them for now to avoid confusion.

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        @Override
+        public SqlRow next() {
```
Shall we add if(closed) check here as well?

devozerov replied:
This method first invokes `hasNext` that does this check.

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlPage.java:
```diff
@@ -0,0 +1,67 @@
+
+import java.util.List;
+import java.util.Objects;
+
```
Javadoc would be useful.

devozerov replied:
Added JavaDoc

petrpleshachkov commented on hazelcast/src/test/java/com/hazelcast/client/txn/ClientTxnUniSocketTest.java:
```diff
@@ -65,6 +65,23 @@ public Member next() {
+                if (members == null || members.length == 0) {
+                    return null;
+                }
+                if (members.length == 2) {
```
Why do we return the last member in this case?

devozerov replied:
I do not know, this is just a copy-paste from the `next()` method to let the test compile. There is this comment above:
`//try to force second member to connect when exist`

devozerov commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java:
```diff
@@ -197,4 +200,14 @@ public static EndpointQualifier createEndpointQualifier(int type, String identif
+        SqlColumnType sqlColumnType = SqlColumnType.getById(type);
+
+        if (sqlColumnType == null) {
+            throw new HazelcastException("Unexpected SQL column type type = [" + type + "]");
```
Fixed

sancar commented:
There seems to be a missing part threading on the server-side. We are marking the message tasks so that they get into the correct thread. See ClientEngineImpl.accept
```
 public void accept(ClientMessage clientMessage) {
        Connection connection = clientMessage.getConnection();
        MessageTask messageTask = messageTaskFactory.create(clientMessage, connection);
        OperationServiceImpl operationService = nodeEngine.getOperationService();
        if (isUrgent(messageTask)) {
            operationService.execute((UrgentMessageTask) messageTask);
        } else if (messageTask instanceof AbstractPartitionMessageTask) {
            operationService.execute((AbstractPartitionMessageTask) messageTask);
        } else if (isQuery(messageTask)) {
            queryExecutor.execute(messageTask);
        } else if (messageTask instanceof TransactionalMessageTask) {
            blockingExecutor.execute(messageTask);
        } else if (messageTask instanceof BlockingMessageTask) {
            blockingExecutor.execute(messageTask);
        } else {
            executor.execute(messageTask);
        }
    }
```
blockingExecutor and queryExecutor run the tasks that do a remote call and wait on them on the thread. 

Any task that runs on `executor` or `operationService` should not wait for a remote call to return a response.They should return the response in an async manner. 

For the new message tasks, it is preferable for them to not block the thread they are running on. In that case, we can leave then on `executor`. If we can not convert them to async, it is better to put them on `queryExecutor`

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1378/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1378/Hazelcast-pr-builder-1378.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1379/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1379/Hazelcast-pr-builder-1379.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:
```diff
@@ -0,0 +1,208 @@
+
+            SqlExecuteCodec.ResponseParameters responseParameters = SqlExecuteCodec.decodeResponse(responseMessage);
+
+            SqlExecuteResponse response = new SqlExecuteResponse(
```
There is no need for `SqlExecuteResponse` object anymore here.

devozerov replied:
Fixed.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:
```diff
@@ -0,0 +1,208 @@
+            ClientMessage requestMessage = SqlFetchCodec.encodeRequest(queryId, cursorBufferSize);
+            ClientMessage responseMessage = invoke(requestMessage, connection);
+            SqlFetchCodec.ResponseParameters responseParameters = SqlFetchCodec.decodeResponse(responseMessage);
+            SqlFetchResponse response = new SqlFetchResponse(responseParameters.rowPage, responseParameters.error);
```
There is no need for `SqlFetchResponse ` object anymore here.

devozerov replied:
Fixed.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1380/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1380/Hazelcast-pr-builder-1380.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java:
```diff
@@ -0,0 +1,178 @@
+
+    private static boolean fetchPage(
+        Iterator<SqlRow> iterator,
+        List<SqlPageRow> page,
```
Would it be possible to use List<Data> instead of SqlPageRow here. And also in the protocol ?  
It would eliminate the need for creating a new type in the protocol, as well as creating extra objects here. 
In the protocol, we may also eliminate SqlPage in the protocol and put the `last boolean` as one of the response params.

devozerov replied:
Removed `SqlPage` from the protocol, although it is still used in the server-side part, because we need to return rows and a flag anyway.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlFetchMessageTask.java:
```diff
@@ -0,0 +1,96 @@
+            error = SqlClientUtils.exceptionToClientError(e, localMemberId);
+        }
+
+        return new SqlFetchResponse(page, error);
```
You can eliminate the need for creating `SqlFetchResponse` class by returning `page` or `error` directly. 
encodeResponse method could act accordingle by checking if `response` is instanceOf Exception or not.

devozerov replied:
It will make the code harder to follow because we will use different ways of returning results for `execute` and `fetch` commands.

sancar commented on hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java:
```diff
@@ -0,0 +1,178 @@
+    private QueryClientState getClientCursor(UUID clientId, QueryId queryId) {
+        QueryClientState cursor = clientCursors.get(queryId);
+
+        if (cursor == null || (clientId != null && !cursor.getClientId().equals(clientId))) {
```
@devozerov This part is confusing to me. Can you document when can a client be null? To me, it looked like it can not, or maybe I am missing something.

devozerov replied:
Fixed. Part of this expression was redundant.

devozerov commented:
@sancar I re-routed tasks to the query executor for now. Previously we decided to proceed with a blocking approach for 4.1 due to lack of time. It will be improved in 4.2+.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1390/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1390/Hazelcast-pr-builder-1390.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1391/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1391/Hazelcast-pr-builder-1391.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1392/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1392/Hazelcast-pr-builder-1392.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1393/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1393/Hazelcast-pr-builder-1393.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

viliam-durina commented on hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java:
```diff
@@ -58,22 +58,72 @@ public QueryId() {
+    public static QueryId parse(String input) {
+        assert input != null;
+
+        String[] parts = input.split("_");
+
+        if (parts.length != 2) {
+            throw new IllegalArgumentException("Query ID is malformed: " + input);
+        }
+
+        try {
+            UUID memberId = UUID.fromString(parts[0]);
+            UUID localId = UUID.fromString(parts[1]);
```
Javadoc of `UUID.toString()` specifies that the result is always 36 chars, so we could just split there. But `UUID.fromString` doesn't assume that, so I don't know. The below code is safe in this regard and doesn't deal with regexes.

```suggestion
        int underscorePos = input.indexOf("_");

        if (underscorePos < 0) {
            throw new IllegalArgumentException("Query ID is malformed: " + input);
        }

        try {
            UUID memberId = UUID.fromString(input.substring(0, underscorePos));
            UUID localId = UUID.fromString(input.substring(underscorePos + 1);
```

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1407/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1407/Hazelcast-pr-builder-1407.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java:[82,77] error: ')' expected
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hazelcast: Compilation failure
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java:[82,77] error: ')' expected
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java:[82,77] error: ')' expected
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1416/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1416/Hazelcast-pr-builder-1416.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

devozerov commented:
HI @asimarslan @sancar, I addressed the concerns we discussed:
1. Implemented special handling for the incompatible messages
1. Added the design doc: https://github.com/hazelcast/hazelcast/blob/baa16db36108a7ed24a49a547ac4e032941fe2b1/docs/design/sql/05-client-support.md

Could you please perform the final review? Also please pay attention to the enterprise part, where the security is implemented: https://github.com/hazelcast/hazelcast-enterprise/pull/3666

Please let me know if you have any additional concerns, or if the design document misses some important information.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1484/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1484/Hazelcast-pr-builder-1484.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientReliableTopicOnClusterRestartTest.shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant:185->lambda$shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant$5:187
[INFO] 
[ERROR] Tests run: 31144, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

devozerov commented:
run-lab-run

asimarslan commented on docs/design/sql/05-client-support.md:
```diff
@@ -0,0 +1,164 @@
+### 2.1 Common Entities
+
+The protocol contains the following common entities that are used across different commands:
+- **Query ID** - an opaque query identifier of type `String` that is used to locate the server-side cursor on the member
```
Can you update this doc?

devozerov replied:
Updated the document (see the last two commits)
